Core/include/Acts/SpacePointFormation/detail/SpacePointBuilder.ipp

```diff
@@ -0,0 +1,673 @@
+    const auto& ends2 = endsOfStrip(gctx, *(mp.second));
+
+    spaPoPa.q = ends1.first - ends1.second;
+    spaPoPa.r = ends2.first - ends2.second;
+
+    const auto& slink_front = getSourceLink(*(mp.first));
+    const auto& slink_back = getSourceLink(*(mp.second));
+
+    boost::container::static_vector<const SourceLink*, 2> slinks;
+    slinks.emplace_back(slink_front);
+    slinks.emplace_back(slink_back);
+
+    double resultPerpProj;
+
+    if (m_config.usePerpProj) {  // for cosmic without vertex constraint
+      resultPerpProj = detail::calcPerpendicularProjection(
+          ends1.first, ends2.first, spaPoPa.q, spaPoPa.r);
+      if (resultPerpProj <= 0.) {
+        Vector3 pos = ends1.first + resultPerpProj * spaPoPa.q;
+        double theta = acos(spaPoPa.q.dot(spaPoPa.r) /
+                            (spaPoPa.q.norm() * spaPoPa.r.norm()));
+        const auto gcov =
+            calcGlobalVars(gctx, *(mp.first), *(mp.second), theta);
+        const double varRho = gcov[0];
+        const double varZ = gcov[1];
+
+        auto sp = spacepoint_t(pos, varRho, varZ, std::move(slinks));
+        spacePointIt = std::move(sp);
+        continue;
+      }
+    }
+
+    bool spFound = calculateDoubleHitSpacePoint(
+        ends1, ends2, m_config.vertex, spaPoPa, m_config.stripLengthTolerance);
+    if (!spFound)
+      spFound =
+          detail::recoverSpacePoint(spaPoPa, m_config.stripLengthGapTolerance);
+
+    if (spFound) {
+      Vector3 pos = 0.5 * (ends1.first + ends1.second + spaPoPa.m * spaPoPa.q);
+
+      double theta = acos(spaPoPa.q.dot(spaPoPa.r) /
+                          (spaPoPa.q.norm() * spaPoPa.r.norm()));
+
+      const auto gcov = calcGlobalVars(gctx, *(mp.first), *(mp.second), theta);
+      const double varRho = gcov[0];
+      const double varZ = gcov[1];
+      auto sp = spacepoint_t(pos, varRho, varZ, std::move(slinks));
+      spacePointIt = std::move(sp);
+    }
+  }
+}
+
+template <typename spacepoint_t>
+std::pair<Vector3, Vector3> SpacePointBuilder<spacepoint_t>::endsOfStrip(
+    const GeometryContext& gctx, const Measurement& measurement) const {
+  Vector3 topGlobal(0, 0, 0);
+  Vector3 bottomGlobal(0, 0, 0);
+
+  auto localPos = getLocalPos(measurement);
+
+  const auto& slink = getSourceLink(measurement);
+  const auto geoId = slink->geometryId();
+  const Surface* surface = m_config.trackingGeometry->findSurface(geoId);
+
+  auto detectorElement = dynamic_cast<const IdentifiedDetectorElement*>(
+      surface->associatedDetectorElement());
+
+  if (!detectorElement) {
+    ACTS_ERROR(" No detector element found in the strip SP formation");
+    return std::make_pair(topGlobal, bottomGlobal);
+  }
+
+  auto digitizationModule = detectorElement->digitizationModule();
+  if (!digitizationModule) {
+    ACTS_ERROR(" No digitizationModule found in the strip SP formation");
+    return std::make_pair(topGlobal, bottomGlobal);
+  }
+  const Segmentation& segment = digitizationModule->segmentation();
```

Comment:
@paulgessinger , do you have an idea to get the segmentation without using IdentifiedDetectorElement?